### PR TITLE
[Merged by Bors] - feat(NumberTheory.NumberField.Basic): add  mem_span_integralBasis

### PR DESCRIPTION
**Pull request by xroblot**
Add the following result: 
```lean
theorem mem_span_integralBasis {x : K} :
    x ∈ Submodule.span ℤ (Set.range (integralBasis K)) ↔ x ∈ 𝓞 K
```
that is, `integralBasis` is indeed a `ℤ`-basis of the ring of integers. 

Co-authored-by: Riccardo Brasca <riccardo.brasca@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
Why not having an actual `Basis ... ℤ (𝓞 K)`? We have somewhere that this induces automatically a `ℚ`-basis of `K` I think.

**Comment by xroblot** [comment]
> Why not having an actual `Basis ... ℤ (𝓞 K)`? We have somewhere that this induces automatically a `ℚ`-basis of `K` I think.

This is what we are doing here. We start with a ℤ-basis of `𝓞 K`, that's `NumberField.RingOfIntegers.basis`, then we prove that it is a ℚ-basis of `K`, that's `NumberField.integralBasis`, and this result closes the circle by proving that the ℤ-lattice generated by this basis inside `K` is indeed `𝓞 K`.

**Comment by riccardobrasca** [comment]
Ah, I see, `NumberField.integralBasis` is defined using `RingOfIntegers.basis`  via `  Basis.localizationLocalization`. Can you state a general result about `Basis.localizationLocalization` instead of this particular case?

**Comment by xroblot** [comment]
> Ah, I see, `NumberField.integralBasis` is defined using `RingOfIntegers.basis` via ` Basis.localizationLocalization`. Can you state a general result about `Basis.localizationLocalization` instead of this particular case?

I added the general result
```lean
 theorem Basis.localizationLocalization_span {ι : Type _} (b : Basis ι R A) :
     Submodule.span R (Set.range (b.localizationLocalization Rₛ S Aₛ)) =
       Submodule.map (IsScalarTower.toAlgHom R A Aₛ) ⊤ 
```
and use it to prove `mem_span_integralBasis`

File: Mathlib/RingTheory/Localization/Module.lean [hunk 1]
```diff
@@ -143,6 +143,17 @@ theorem Basis.localizationLocalization_repr_algebraMap {ι : Type _} (b : Basis
     _ = algebraMap R Rₛ (b.repr x i) := by simp [Algebra.smul_def]
 #align basis.localization_localization_repr_algebra_map Basis.localizationLocalization_repr_algebraMap
 
+theorem Basis.localizationLocalization_span {ι : Type _} (b : Basis ι R A) :
+    Submodule.span R (Set.range (b.localizationLocalization Rₛ S Aₛ)) =
+      Submodule.map (IsScalarTower.toAlgHom R A Aₛ) ⊤ := by
+  convert congrArg (Submodule.map (IsScalarTower.toAlgHom R A Aₛ)) b.span_eq
+  change _ = map (AlgHom.toLinearMap _) _
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Localization/Module.lean:150]: Can you try to use
```
{F : Type _} [SemilinearMapClass F σ₁₂ M M₂]
```
and `(f : F)` in `Submodule.map_span`? This should make the `change` line useless. I see that most of `Mathlib.LinearAlgebra.Span` already uses `SemilinearMapClass`, it should be a pretty easy generalization.

Comment (xroblot) [reply Mathlib/RingTheory/Localization/Module.lean:150]: I am sorry, but I cannot see what you want me to do. I tried to remove the `change` and provide some explicit arguments to `Submodule.map_span` but I could not get anywhere without making things more complicated

Comment (riccardobrasca) [reply Mathlib/RingTheory/Localization/Module.lean:150]: I've pushed a commit, let's see what CI think thinks. If it becomes too complicated feel free to revert it.

Comment (riccardobrasca) [reply Mathlib/RingTheory/Localization/Module.lean:150]: The point is that `Submodule.map_span` is stated for `f : M →ₛₗ[σ₁₂] M₂` but it actually holds for any `(f : F)`, where `[SemilinearMapClass F σ₁₂ M M₂]`. You have `f : A →ₐ[R] Aₛ`, that is not literally a semilinear map, so you need the `change` line. But we have `[SemilinearMapClass (A →ₐ[R] Aₛ) ... ...]`, so with this generalization the `change` line should become useless.

In general, when a `change` is needed I see this as a warning that a lemma is missing (in this case we can prove, by `rfl`, that `Submodule.map f M = Submodule.map f.toLinearMap M` and rewrite this) or that something is stated in a suboptimal way.

Comment (xroblot) [reply Mathlib/RingTheory/Localization/Module.lean:150]: Ok. I see. Thanks for the explanation!

Comment (riccardobrasca) [reply Mathlib/RingTheory/Localization/Module.lean:150]: The reason why rewriting the `rfl` lemma is better than using `change` is the following: suppose that the `change` trick is used 100 times and I refactor the definition of `Submodule.map` in such a way that the `Submodule.map f M = Submodule.map f.toLinearMap M` is not true definitionally anymore (this seems unlike, I agree, but we changed the definition of groups two years ago...). Then I need to fix all the 100 occurrences of the trick, by hand. If instead `rw` is used, I only need to fix it once, in the proof of the lemma.

**Comment by riccardobrasca** [comment]
@Vierkantor can you have a look at this? Thanks!

**Comment by Vierkantor** [comment]
Sure! I will make some time in the coming days.

**Comment by riccardobrasca** [comment]
Not sure why "Detect changes to header SHAs / Add annotations (pull_request) " is taking forever...

**Comment by riccardobrasca** [comment]
> Not sure why "Detect changes to header SHAs / Add annotations (pull_request) " is taking forever...

It seems merging master solved this.

**Comment by xroblot** [comment]
> > Not sure why "Detect changes to header SHAs / Add annotations (pull_request) " is taking forever...
> 
> It seems merging master solved this.

Riccardo, thanks a lot for your improvements of this PR. I am in Japan at the moment and it is difficult to find time to work on Lean. I added your name as a co-author. Maybe I should also update the title and description?

File: Mathlib/RingTheory/Localization/Module.lean [hunk 2]
```diff
@@ -140,6 +140,16 @@ theorem Basis.localizationLocalization_repr_algebraMap {ι : Type _} (b : Basis
...
 
+theorem Basis.localizationLocalization_span {ι : Type _} (b : Basis ι R A) :
+    Submodule.span R (Set.range (b.localizationLocalization Rₛ S Aₛ)) =
+      Submodule.map (IsScalarTower.toAlgHom R A Aₛ) ⊤ := by
+  convert congrArg (Submodule.map (IsScalarTower.toAlgHom R A Aₛ)) b.span_eq
+  rw [Submodule.map_span]
+  congr
+  ext
+  simp only [Set.mem_range, localizationLocalization_apply, Set.mem_image, exists_exists_eq_and]
+  rfl
```
Comment (Vierkantor) [inline Mathlib/RingTheory/Localization/Module.lean:151]: The right hand side can be simplified further using `Submodule.map_top`.

Also, I found the `convert`s and `rw`s hard to read, how about `calc`?
```suggestion
theorem Basis.localizationLocalization_span {ι : Type _} (b : Basis ι R A) :
    Submodule.span R (Set.range (b.localizationLocalization Rₛ S Aₛ)) =
      LinearMap.range (IsScalarTower.toAlgHom R A Aₛ) :=
  calc span R (Set.range ↑(localizationLocalization Rₛ S Aₛ b))
    _ = span R (↑(IsScalarTower.toAlgHom R A Aₛ) '' Set.range ↑b) := by congr; ext; simp
    _ = map (IsScalarTower.toAlgHom R A Aₛ) (span R (Set.range b)) := by rw [Submodule.map_span]
    _ = LinearMap.range (IsScalarTower.toAlgHom R A Aₛ) := by rw [b.span_eq, Submodule.map_top]
```

File: Mathlib/NumberTheory/NumberField/Basic.lean [hunk 3]
```diff
@@ -168,6 +168,12 @@ theorem integralBasis_apply (i : Free.ChooseBasisIndex ℤ (𝓞 K)) :
   Basis.localizationLocalization_apply ℚ (nonZeroDivisors ℤ) K (RingOfIntegers.basis K) i
 #align number_field.integral_basis_apply NumberField.integralBasis_apply
 
+theorem mem_span_integralBasis {x : K} :
+    x ∈ Submodule.span ℤ (Set.range (integralBasis K)) ↔ x ∈ 𝓞 K := by
+  rw [integralBasis, Basis.localizationLocalization_span, ← SetLike.mem_coe, ← SetLike.mem_coe,
+    Submodule.map_coe, IsScalarTower.coe_toAlgHom', Submodule.top_coe, Set.image_univ,
+    show Set.range _ = Set.range ((↑) : (𝓞 K) → K) by rfl, Subtype.range_coe]
```
Comment (Vierkantor) [inline Mathlib/NumberTheory/NumberField/Basic.lean:175]: Let's split this into two parts: showing the output of `localizationLocalization_span` is actually the algebra, and then connecting this with `integralBasis`:
```suggestion
-- This should be moved, perhaps to Mathlib.Algebra.Algebra.Subalgebra.Tower?
@[simp]
lemma _root_.Subalgebra.range_isScalarTower_toAlgHom {R A} [CommSemiring R] [CommSemiring A]
    [Algebra R A] (S : Subalgebra R A) :
    LinearMap.range (IsScalarTower.toAlgHom R S A) = Subalgebra.toSubmodule S := by
  ext
  simp only [← Submodule.range_subtype (Subalgebra.toSubmodule S), LinearMap.mem_range,
    IsScalarTower.coe_toAlgHom', Subalgebra.mem_toSubmodule]
  rfl

theorem mem_span_integralBasis {x : K} :
    x ∈ Submodule.span ℤ (Set.range (integralBasis K)) ↔ x ∈ 𝓞 K := by
  rw [integralBasis, Basis.localizationLocalization_span, Subalgebra.range_isScalarTower_toAlgHom,
      Subalgebra.mem_toSubmodule]
```

**Review by Vierkantor — COMMENTED** [review]
This is indeed a useful result, thanks for spotting that we're missing it.

**Comment by xroblot** [comment]
@Vierkantor Thanks for the review. As you suggested, I moved `Subalgebra.range_isScalarTower_toAlgHom` to `Mathlib.Algebra.Algebra.Subalgebra.Tower`

**Review by riccardobrasca — APPROVED** [review]
LGTM, thanks!

**Review by Vierkantor — APPROVED** [review]
Thanks!

bors r+

**Pull request closed without merging**